src/fft.rg

```diff
@@ -13,74 +13,135 @@
 fftw_c.FFTW_ESTIMATE = (2 ^ 6)
-fftw_c.FFTW_WISDOM_ONLY = (2 ^ 21)
 
 local fft = {}
 
-function fft.generate_fft_interface(itype, dtype)
+--- Create the FFT interface.
+-- @param itype Index type of transform (int2d/int2d/int3d/int4d).
+-- @param dtype_in Input data type of transform (float/double/complex32/complex64).
+-- @param dtype_out Output data type of transform (complex32/complex64).
+function fft.generate_fft_interface(itype, dtype_in, dtype_out)
   assert(regentlib.is_index_type(itype), "requires an index type as the first argument")
   local dim = itype.dim
-  assert(dim >= 1 and dim <= 3, "currently only 1 <= dim <= 3 is supported")
-  assert(dtype == complex64, "currently only complex64 is supported")
+  local dtype_out_size = terralib.sizeof(dtype_out)
+  assert(dim >= 1 and dim <= 4, "currently only 1 <= dim <= 4 is supported")
+  assert(dtype_in == float or dtype_in == double or dtype_in == complex32 or dtype_in == complex64, "input type must be float/double/complex32/complex64")
+  assert(dtype_out == complex32 or dtype_out == complex64, "output type must be complex32/complex64")
+
+   -- Single-precision transforms
+  local float_to_complex32_transform = (dtype_in == float and dtype_out == complex32)
+  local complex32_to_complex32_transform = (dtype_in == complex32 and dtype_out == complex32)
+
+   -- Double-precision transforms
+  local double_to_complex64_transform = (dtype_in == double and dtype_out == complex64)
+  local complex64_to_complex64_transform = (dtype_in == complex64 and dtype_out == complex64)
 
   local iface = {}
 
+  -- Create fspaces depending on whether GPUs are used or not
   local iface_plan
-  if use_cuda then
+  if gpu_available then
     fspace iface_plan {
       p : fftw_c.fftw_plan,
+      float_p : fftw_c.fftwf_plan,
       cufft_p : cufft_c.cufftHandle,
       address_space : c.legion_address_space_t,
     }
   else
     fspace iface_plan {
       p : fftw_c.fftw_plan,
+      float_p : fftw_c.fftwf_plan,
       address_space : c.legion_address_space_t,
     }
   end
+
   iface.plan = iface_plan
-  iface.plan.__no_field_slicing = true -- don't field slice this struct
+  iface.plan.__no_field_slicing = true -- Don't field slice this struct
 
   local function make_get_base(d, t)
     local rect_t = c["legion_rect_" .. d .. "d_t"]
     local get_accessor = c["legion_physical_region_get_field_accessor_array_" .. d .. "d"]
     local raw_rect_ptr = c["legion_accessor_array_" .. d .. "d_raw_rect_ptr"]
     local destroy_accessor = c["legion_accessor_array_" .. d .. "d_destroy"]
 
+    local struct base_pointer_t {
+      base : &t,
+      offset : c.legion_byte_offset_t[d],
+      dtype_size : int
+    }
+
     local terra get_base(rect : rect_t,
-                 physical : c.legion_physical_region_t,
-                 field : c.legion_field_id_t)
+                physical : c.legion_physical_region_t,
+                field : c.legion_field_id_t)
```

Comment:
Hard to see in the PR interface here, but I would line this up with the start of the `rect` parameter. (If you want to fix the spacing. Or else I'd remove the diff entirely and let it go back to the original, which is still wrong, but at least doesn't cause a diff for something you're not fixing anyway.)

Reply:
Lined this up.